checker/src/org/checkerframework/checker/lock/LockVisitor.java

```diff
@@ -4,88 +4,340 @@
+
+        SideEffectAnnotation sea = atypeFactory.methodSideEffectAnnotation(methodElement, true);
+
+        if (sea == SideEffectAnnotation.MAYRELEASELOCKS) {
+            boolean issueGSwithMRLWarning = false;
+
+            VariableTree receiver = node.getReceiverParameter();
+            if (receiver != null) {
+                if (atypeFactory.getAnnotatedType(receiver).hasAnnotation(checkerGuardSatisfiedClass)) {
+                    issueGSwithMRLWarning = true;
+                }
+            }
+
+            if (!issueGSwithMRLWarning) { // Skip this loop if it is already known that the warning must be issued.
+                for(VariableTree vt : node.getParameters()) {
+                    if (atypeFactory.getAnnotatedType(vt).hasAnnotation(checkerGuardSatisfiedClass)) {
+                        issueGSwithMRLWarning = true;
+                        break;
+                    }
+                }
+            }
+
+            if (issueGSwithMRLWarning) {
+                checker.report(Result.failure("guardsatisfied.with.mayreleaselocks"), node);
+            }
+        }
+
+        // Issue an error if a non-constructor method definition has a return type of @GuardSatisfied without an index.
+        if (methodElement != null && methodElement.getKind() != ElementKind.CONSTRUCTOR) {
+	        AnnotatedTypeMirror returnTypeATM = atypeFactory.getAnnotatedType(node).getReturnType();
+
+	        if (returnTypeATM != null && returnTypeATM.hasAnnotation(GuardSatisfied.class)) {
+	            int returnGuardSatisfiedIndex = AnnotationUtils.
+	                    getElementValue(returnTypeATM.getAnnotation(GuardSatisfied.class), "value", Integer.class, true);
+
+	            if (returnGuardSatisfiedIndex == -1) {
+	                checker.report(Result.failure("guardsatisfied.return.must.have.index"), node);
+	            }
+	        }
+        }
+
+        return super.visitMethod(node, p);
+    }
+
+    // When visiting a method call, if the receiver formal parameter
+    // has type @GuardSatisfied, skip the receiver subtype check and instead
+    // verify that the guard is satisfied.
+    @Override
+    protected boolean skipReceiverSubtypeCheck(MethodInvocationTree node,
+            AnnotatedTypeMirror methodDefinitionReceiver,
+            AnnotatedTypeMirror methodCallReceiver) {
+        Set<AnnotationMirror> annos = methodDefinitionReceiver.getAnnotations();
+        for(AnnotationMirror anno : annos) {
+            if (AnnotationUtils.areSameByClass(anno, checkerGuardSatisfiedClass)) {
+                Element invokedElement = TreeUtils.elementFromUse(node);
+
+                boolean receiverIsThatOfEnclosingMethod = false;
+
+                MethodInvocationNode nodeNode = (MethodInvocationNode) atypeFactory.getNodeForTree(node);
+
+                Node receiverNode = nodeNode.getTarget().getReceiver();
+                if (receiverNode instanceof ExplicitThisLiteralNode ||
```

Comment:
ExplicitThisLiteralNode and ImplicitThisLiteralNode are subtypes of ThisLiteralNode. Why isn't the instanceof check for the supertype enough?